// serverio/lib/http-copy.js
var connect = require('connect'),
    http = require("http")
    // ,session = require('express-session')
    ;

function rinfo(tcpstream, data) {
    this.address = tcpstream.remoteAddress;
    this.port = tcpstream.remotePort;
    this.family = tcpstream.address().family;
    this.size = data.length;
}

exports.start = function(config, callback){
    // var token = config.http_token;

    var app = connect()
    .use(connect.favicon())
    .use(connect.cookieParser())
    .use(connect.session({ secret: 'secretSessionWordGoesHere', cookie: { maxAge: /*60000*/ 600000 }}))
    .use(config.endpoint || '/metrix', function (req, res, next) {

        req.setEncoding('ascii');

        // console.log(req.method, req.originalUrl, '  - session:', req.sessionID);
        (req.session.views && req.session.views++) || (req.session.views = 1);
        // console.log('session: ', req.session.views);

        var buffer = '';
        req.on('data', function (data) {
            buffer += data;
        });
        req.on('end', function (data) {

            console.log('data: ', data, 'buffer:', buffer);

            if (data) { buffer += data; }
            callback(buffer, new rinfo(req.socket, buffer));
            // res.statusCode = 401; res.end('error..');
            res.end('');
        });

    });

    http.createServer(app).listen(parseInt(config.port || 8080, config.address || undefined, 10));
    // server.listen(config.port || 8080, config.address || undefined);
    return true;
};

// exports.start = function(config, callback){
//     var token = config.http_token;
//
//     var server = http.createServer(function (request, response) {
//         // Check username/password
//         if (token && !(request.headers.token === token)) {
//             response.statusCode = 401;
//             response.end('Invalid token');
//             return;
//         }
//
//         request.setEncoding('ascii');
//
//         var buffer = '';
//         request.on('data', function (data) {
//             buffer += data;
//         });
//         request.on('end', function (data) {
//             if (data) { buffer += data; }
//             callback(buffer, new rinfo(request.socket, buffer));
//             response.end('');
//         });
//     });
//
//     server.listen(config.port || 8080, config.address || undefined);
//     return true;
// };
